project/TestParallelization.scala

```diff
@@ -0,0 +1,106 @@
+import sbt.Keys._
+import sbt._
+
+object TestParallelization {
+
+  lazy val testGroupingInDifferentJVMSettings = Seq(
+    Test / testGrouping := {
+      val tests = (Test / definedTests).value
+      val groupingStrategy = (Test / testGroupingStrategy).value
+      val grouping = tests.foldLeft(groupingStrategy) {
+        case (strategy, testDefinition) => strategy.add(testDefinition)
+      }
+      val logger = streams.value.log
+      logger.info(s"Tests will be grouped in ${grouping.testGroups.size} groups")
+      val groups = grouping.testGroups
+      groups.foreach{
+          group =>
+            logger.info(s"${group.name} contains ${group.tests.size} tests")
+      }
+      groups
+    }
+  )
+
+  lazy val simpleGroupingStrategySettings = Seq(
+    Test / forkTestJVMCount := {
+      sys.env.get("DELTA_TEST_JVM_COUNT").map(_.toInt).getOrElse(4)
+    },
+    Test / testGroupingStrategy := {
+      val groupsCount = (Test / forkTestJVMCount).value
+      val baseJvmDir = baseDirectory.value
+      SimpleHashStrategy(groupsCount, baseJvmDir, defaultForkOptions.value)
+    },
+    Test / parallelExecution := true,
+    Global / concurrentRestrictions := {
+      Seq(Tags.limit(Tags.ForkedTestGroup, (Test / forkTestJVMCount).value))
+    }
+  )
+
+
+  val forkTestJVMCount = SettingKey[Int]("fork test jvm count",
+    "The number of separate JVM to use for tests"
+  )
+
+  val testGroupingStrategy = TaskKey[GroupingStrategy]("test grouping strategy",
+    "The strategy to allocate different groups to different jvms"
+  )
+  private val defaultForkOptions = Def.task {
+    ForkOptions(
+      javaHome = javaHome.value,
+      outputStrategy = outputStrategy.value,
+      bootJars = Vector.empty,
+      workingDirectory = Some(baseDirectory.value),
+      runJVMOptions = (Test / javaOptions).value.toVector,
+      connectInput = connectInput.value,
+      envVars = (Test / envVars).value
+    )
+  }
+
+  sealed trait GroupingStrategy {
+
+    def add(testDefinition: TestDefinition): GroupingStrategy
+
+    def testGroups: List[Tests.Group]
+
+  }
+
+  class SimpleHashStrategy(groups: Map[Int, Tests.Group]) extends GroupingStrategy {
```

Comment:
minor: make the constructor private so that callers are forced to use the apply().

Reply:
💯 thank you for catching that